vm/systemSmartContracts/eei.go

```diff
@@ -310,6 +327,71 @@ func (host *vmContext) mergeContext(currContext *vmContext) {
 	host.scAddress = currContext.scAddress
 }
 
+func (host *vmContext) properMergeContexts(parentContext *vmContext, returnCode vmcommon.ReturnCode) {
+	if !host.enableEpochsHandler.IsMultiClaimOnDelegationEnabled() {
+		host.mergeContext(parentContext)
+		return
+	}
+
+	host.scAddress = parentContext.scAddress
+	host.AddReturnMessage(parentContext.returnMessage)
+	if returnCode != vmcommon.Ok {
+		// no need to merge - revert was done - transaction will fail
+		return
+	}
+
+	host.output = append(host.output, parentContext.output...)
+	for _, rightAccount := range parentContext.outputAccounts {
+		leftAccount, exist := host.outputAccounts[string(rightAccount.Address)]
+		if !exist {
+			leftAccount = &vmcommon.OutputAccount{
+				Balance:      big.NewInt(0),
+				BalanceDelta: big.NewInt(0),
+				Address:      rightAccount.Address,
+			}
+			host.outputAccounts[string(rightAccount.Address)] = leftAccount
+		}
+		addOutputAccounts(leftAccount, rightAccount)
+	}
+}
+
+func addOutputAccounts(
+	destination *vmcommon.OutputAccount,
+	rightAccount *vmcommon.OutputAccount,
+) {
+	if len(rightAccount.Address) != 0 {
+		destination.Address = rightAccount.Address
+	}
+	if rightAccount.Balance != nil {
+		destination.Balance = rightAccount.Balance
+	}
+	if destination.BalanceDelta == nil {
```

Comment:
can we have nil here? The left account on L347 will always have the BalanceDelta set.

Reply:
Only for extra protection